### PR TITLE
#47654 Adds a new `constraints` parameter to pip command

### LINKED ISSUE #47654: pip: Add "constraints" value
##### SUMMARY
Adds support for the --constraint <file> argument to `pip` command

##### ISSUE TYPE
- Feature Idea

##### COMPONENT NAME
pip module

##### ADDITIONAL INFORMATION
Adds the parameter "--constraint <file>", which permit to use an additional file to set some contraints, and also make requirement file more readable.
```yaml
- name: Installs python requirements
  pip:
    virtualenv_python: " /usr/bin/python3"
    virtualenv: "{{ venv_path }}"
    requirements: "{{ requirement_file }}"
    constraints: "{{ constraint_file }}"
    ...
```

### PR DESCRIPTION
**Pull request by frague59**
##### SUMMARY
Adds the `constraints` argument to the pip packaging command

Fixes #47654

##### ISSUE TYPE
- Feature Pull Request

##### COMPONENT NAME
pip

##### ADDITIONAL INFORMATION
Adds a new parameter for the pip command: constraints

```yml
- name: Installs requirements with constraints
  pip:
    requirements: requirements.txt
    constraints: constraints.txt
```

### CODE REVIEW

File: lib/ansible/modules/packaging/language/pip.py [hunk 1]
```diff
@@ -5,6 +5,21 @@
...
+import shlex
+import sys
+import tempfile
+import traceback
+from distutils.version import LooseVersion
+
+from ansible.module_utils._text import to_native
+from ansible.module_utils.basic import (AnsibleModule, is_executable,
+                                        missing_required_lib)
+from ansible.module_utils.six import PY3
```
Comment (sivel) [inline lib/ansible/modules/packaging/language/pip.py:21]: These imports need to be moved back to where they were.  As you can see moving it has caused CI to fail.

File: lib/ansible/modules/packaging/language/pip.py [hunk 2]
```diff
@@ -31,6 +46,10 @@
     description:
       - The path to a pip requirements file, which should be local to the remote system.
         File can be specified as a relative path if using the chdir option.
+  constraints:
+    description:
+      - The path to a pip constraints file, which should be local to the remote system.
+        File can be specified as a relative path if using the chdir option.
```
Comment (sivel) [inline lib/ansible/modules/packaging/language/pip.py:52]: Need `version_added: "2.8"` here

File: lib/ansible/modules/packaging/language/pip.py [hunk 3]
```diff
@@ -567,6 +585,7 @@ def main():
             name=dict(type='list'),
             version=dict(type='str'),
             requirements=dict(type='str'),
+            constraints=dict(type='str'),
```
Comment (sivel) [inline lib/ansible/modules/packaging/language/pip.py:588]: arguments pointing to a file should use `type='path'`.

File: lib/ansible/modules/packaging/language/pip.py [hunk 4]
```diff
@@ -682,6 +702,8 @@ def main():
             cmd.extend(to_native(p) for p in packages)
         elif requirements:
             cmd.extend(['-r', requirements])
+            if constraints:
```
Comment (sivel) [inline lib/ansible/modules/packaging/language/pip.py:705]: constraints isn't dependent on requirements.  constraints can still be useful if the libraries being installed are liberal in their selection.

This should be moved outside.

File: lib/ansible/modules/packaging/language/pip.py [hunk 5]
```diff
@@ -743,9 +765,15 @@ def main():
 
         changed = changed or venv_created
 
-        module.exit_json(changed=changed, cmd=cmd, name=name, version=version,
-                         state=state, requirements=requirements, virtualenv=env,
-                         stdout=out, stderr=err)
+        if constraints:
```
Comment (sivel) [inline lib/ansible/modules/packaging/language/pip.py:768]: There is no reason to make this conditional.  We don't do this for `requirements` don't do it here too.  Just use a single `module.exit_json`

**Review by sivel — CHANGES_REQUESTED** [review]
In addition to other changes listed here, please:

1. rebase to remove the merge commits
1. add tests for this functionality
1. add a changelog fragment to `changelog/fragments`

**Comment by samdoran** [comment]
Closing due to inactivity.

**Pull request closed without merging**